gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -252,4 +252,55 @@ MacroBuiltin::compile_error (Location invoc_locus, AST::MacroInvocData &invoc)
   return AST::ASTFragment::create_error ();
 }
 
+AST::ASTFragment
+MacroBuiltin::include (Location invoc_locus, AST::MacroInvocData &invoc)
+{
+  /* Get target filename from the macro invocation, which is treated as a path
+     relative to the include!-ing file (currently being compiled).  */
+  auto lit_expr
+    = parse_single_string_literal (invoc.get_delim_tok_tree (), invoc_locus);
+  if (lit_expr == nullptr)
+    return AST::ASTFragment::create_error ();
+
+  std::string target_filename
+    = source_relative_path (lit_expr->as_string (), invoc_locus);
+
+  RAIIFile target_file (target_filename.c_str ());
+  Linemap *linemap = Session::get_instance ().linemap;
+
+  if (target_file.get_raw () == nullptr)
+    {
+      rust_error_at (lit_expr->get_locus (), "cannot open included file %s: %m",
+		     target_filename.c_str ());
+      return AST::ASTFragment::create_error ();
+    }
+
+  rust_debug ("Attempting to parse included file %s", target_filename.c_str ());
+
+  Lexer lex (target_filename.c_str (), std::move (target_file), linemap);
```

Comment:
`target_filename.c_str ()`

I did not find a way to correctly move `target_filename`'s ownership to either `Lexer` or `Parser`.

This creates an issue where if there are warnings in an `include`d file, the diagnostic information will be garbage data.

Reply:
Yes, this is an issue we also faced when started fetching items from modules. The solution was to keep the string as a member of the `Module` class, which we cannot easily do here. If you can't figure it out, let me know and we'll try to work something out together (for example by adding a `private_data` field to the macro-invocation or something, even though I'm not a big fan of that haha).

Reply:
> Yes, this is an issue we also faced when started fetching items from modules. The solution was to keep the string as a member of the `Module` class, which we cannot easily do here. 

I saw that, that's why I am kind of in a dilemma at the moment.

> If you can't figure it out, let me know and we'll try to work something out together (for example by adding a `private_data` field to the macro-invocation or something, even though I'm not a big fan of that haha).

Yes, that would make the type conversation a bit unsafe when casting data types from the `private_data` pointer.